Create semver 0.x.x guideline (#4)
frequenz-floss/docs: python/semver-0.x.x.md

# Semantic version rules and guidelines for 0.x.x releases

We follow [semantic versioning](https://semver.org/), but we also have many

> projects at 0.x.x, which means no rules at all in semver. So far the convention

Reviewer: `which means no rules at all in semver`

I know what you intend here, but this is not entirely true. There is at least one guideline about major version 0:
https://semver.org/spec/v2.0.0.html#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase

So I would suggest tweaking this line to something like the following (please feel free to rephrase):
```
which means the standard versioning standards are not as rigorous as those for major version > 0.
```

Author: I would argue is it true, there are [no rules](https://semver.org/spec/v2.0.0.html#spec-item-4):

> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

**Anything MAY change at any time** is pretty much no rules IMHO :laughing: 

And this is only a suggestion, not a rule:

> How should I deal with revisions in the 0.y.z initial development phase?
> 
> The simplest thing to do is start your initial development release at 0.1.0 and then increment the minor version for each subsequent release.

So I don't think it changes the "no rules" statement.

About:

> which means the standard versioning standards are not as rigorous as those for major version > 0.

I really think this is more misleading, because it suggest there is any rule to follow for major version 0. But is not that rules for major 0 are less rigorous than version > 0. There are really no rules: **Anything MAY change at any time**.

was to keep patch release as bug fix releases (so 0.X.y), and minor releases

> could include new features and/or breaking changes.

Reviewer: This line needs fixing.

In `frequenz-api-common` and `frequenz-api-microgrid`, we were always following the following rules:
- Update patch version for bug fixes and new features
- Update minor version for breaking changes

Author: This is only for Python.

As we have bigger projects with many dependencies, it is becoming more
challenging to update dependencies when we don't have any guarantees for 0.x.x
versions. As we don't want to prematurely declare a library 1.0.0 (because that
carries some expectations about completeness, not only about breaking changes),
we need to agree on some guarantees for 0.x.x releases.

The goal is to allow developers to move their libraries that are under heavy
development fast, and yet make the life of users of those libraries easier,
without ending in [dependency
hell](https://en.wikipedia.org/wiki/Dependency_hell).

In this document we use the words like MUST, MAY, etc. with the meaning defined
by [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).

## Patch versions guarantees

Provide the same guarantees as minor versions after 1.0.0, paraphrasing semver

> on [minor versions](https://semver.org/#spec-item-7), patch version Z (0.y.Z):

Reviewer: This line reads a bit odd. May I suggest the following?

```suggestion
Patch versions should provide the same guarantees as [minor versions after 1.0.0](https://semver.org/#spec-item-7). Patch version Z (0.y.Z):
```

Author: Fixed, I used "Patch versions MUST provide..." to keep the wording more clear. Also fixed it for Minor releases.

* MUST be incremented if new, backward compatible functionality is introduced
  to the public API.
* MUST be incremented if any public API functionality is marked as deprecated.
* MAY be incremented if substantial new functionality or improvements are
  introduced within the private code.
* MAY include [patch level changes](https://semver.org/#spec-item-6):

  - MUST be incremented if only backward compatible bug fixes are introduced.

>   - A bug fix is defined as an internal change that fixes incorrect behavior.

Reviewer: This part reads odd - MUST under a MAY spec. I would suggest specifying that patch version must be increased on bug fixes.

```suggestion
* MUST be incremented if bug fixes are introduced (a bug fix is defined as an internal change that fixes incorrect behavior).
```

Author: I think the `MUST` inside `MAY` is OK, is like a conditional `IF include patch level changes THEN MUST ...`. 

he rationale for this format is I'm copy&pasting the semver text here, I thought leaving it more verbatim could make things more clear if you are familiar with semver, or if you go read the semver specs while reading these specs. As I envisioned this as an extension of semver, I thought it made sense.

> 6. Patch version Z (x.y.Z | x > 0) MUST be incremented if only backward compatible bug fixes are introduced. A bug fix is defined as an internal change that fixes incorrect behavior.
>
> 7. Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backward compatible functionality is introduced to the public API. It MUST be incremented if any public API functionality is marked as deprecated. It MAY be incremented if substantial new functionality or improvements are introduced within the private code. It MAY include patch level changes. Patch version MUST be reset to 0 when minor version is incremented.
> 
> 8. Major version X (X.y.z | X > 0) MUST be incremented if any backward incompatible changes are introduced to the public API. It MAY also include minor and patch level changes. Patch and minor versions MUST be reset to 0 when major version is incremented.

But if it is more confusing to you, we can reconsider.

Reviewer: The If-MUST phrasing is much more obvious here. The MAY-MUST phrasing is indeed confusing.

Author: What about:

```suggestion
* MAY include *patch level changes*. *Patch level changes* are [defined by semver](https://semver.org/#spec-item-6) as:

  - MUST be incremented if only backward compatible bug fixes are introduced.
  - A bug fix is defined as an internal change that fixes incorrect behavior.
```

Author: I really think changing the verbatim contents of the semver definition would be more confusion from how the document is structured. I'm putting the verbatim definition, and then I separate other extra rules in "Additionally:". I think this makes it very easy to have a mental diff it you are comparing the rules here and the rules in semver. If we don't copy the semver definition verbatim in the items above, I think it doesn't make sense to separate the "Additionally:", so I would need to change the structure of the document quite a bit to make it more consistent.

Because of this I would try hard to keep the original wording of semver in the items above, and only add some small extra clarification if the sub-item is confusing without more context.

Reviewer: You suggested phrasing is a definite improvement over the previous one. It is much clearer. :+1:

Author: Updated.

Additionally:

* Deprecated symbols MAY NOT be maintained, it is recommended to switch to
  non-deprecated symbols as soon as possible.

## Minor versions guarantees

Provide the same guarantees as major versions after 1.0.0, paraphrasing semver

> on [major versions](https://semver.org/#spec-item-8), minor version X (0.Y.z):

Reviewer 2: Seems like there is a typo. Should be X (0.X.z):

Author: There is a typo, but it should be `version Y (0.Y.z)` to follow semver lingo.

Author: Fixed.

> on [major versions](https://semver.org/#spec-item-8), minor version X (0.Y.z):

Reviewer: This line reads a bit odd. May I suggest the following?

```suggestion
Minor versions should provide the same guarantees as [major versions after 1.0.0](https://semver.org/#spec-item-8). Patch version Y (0.Y.z):
```

Author: Fixed

* MUST be incremented if any backward incompatible changes are introduced to
  the public API.
* MAY also include patch level changes.

> * The patch versions MUST be reset to 0 when minor version is incremented.

Reviewer: ```suggestion
* The patch version MUST be reset to 0 when minor version is incremented.
```

Author: Fixed

Additionally:

* Code changes SHOULD not be required if upgrading from the last patch version

>   of the previous minor and no deprecated symbols are used.

Reviewer: Are you sure we will be able to follow this rule?
E.g., if we were to do this in `frequenz-api-common`, then the `v1` package will be marked deprecated in the last patch release. The minor release that will follow will need the older `v1alphaX` package to become `v1`. In this case, code changes will accompany the migration to the minor version.

Author: This is Python only, and it is a SHOULD, so it is recommended but not mandatory:

> 3. **SHOULD**   This word, or the adjective "RECOMMENDED", mean that there
>    may exist valid reasons in particular circumstances to ignore a
>    particular item, but the full implications must be understood and
>    carefully weighed before choosing a different course.

https://datatracker.ietf.org/doc/html/rfc2119

## Library Developer Guidelines

When libraries are still under heavy development but have many users, breaking
releases should be avoided as much as possible, always make patch releases

> where possible.

Reviewer: Nitpick - a minor tweak
```suggestion
When libraries are still under heavy development but have many users, breaking
releases should be avoided as much as possible. Library maintainer are also
recommended to make patch releases whenever possible. 
```

Author: I'll try to improve the wording, but your suggestion for me sounds like it is recommended on any circumstances to make patch releases. What I want to say is if the library is already heavily used by many users, patch releases should be done unless it is really really hard to move forward without a breaking change.

Feeding all of this to Gemini, it suggested:

> When libraries are still under heavy development but have many users, it is highly recommended to prioritize patch releases. Introducing a breaking (minor) release is best reserved for situations where it is unfeasible to move the project forward in a backward-compatible way.

I will apply this, but feel free to follow-up if you think it is not addressing your concerns.

> where possible.

Reviewer: Does it make sense to explicitly mention the following here?
```
Maintainers should try to introduce all planned non-breaking changes in a patch
release before making the next breaking minor release. That way, library users
get access to new features without having to go through a breaking API change,
allowing for a smoother upgrade path.
```

Author: Maybe, I think this is implied by the recommendations that follow (which basically limit all breaking changes in a major release to removing deprecated symbols). Not sure if you made this comment before you read them. Let me know if you still think it would add value and I will include it.

Reviewer: I read the lines below, that is why I am not sure if this needs to be explicitly mentioned in beginning or not, since that could lead to bloat.

I do not have a strong feeling for this one. So if you think it is not needed, we can just resolve this discussion.

Author: OK, if we are unsure I prefer to keep it as it is, this doesn't have to be the end version, we can always update it in the future as we gain more experience.

### Patch releases

To be able to do a patch release instead of a minor release (make the release
backwards compatible):

* Avoid removing symbols, deprecate them instead.

* Mark any unmaintained symbol as `deprecated` as soon as possible to encourage

>   users to update as soon as possible too.

Reviewer: I would suggest incorporating the following:
In protobuf, there is no explicit way of marking symbols as deprecated. So the deprecation notice has to be added to the symbol documentation instead.

Reviewer: This is python only, so this can be resolved.

Author: Python only. We could either extend it to be more general or create another guide for protobuf, but again, I think protobuf have different problems to address than python, so it probably makes more sense to have a separate guide for APIs. Also most of this is handled in protobuf by creating a new path prefix `v1alphaX` each time we introduce breaking changes, so it is not a big issue there, and the approach is simpler I think.
